<a href="https://colab.research.google.com/github/Rick1310/DS-Unit-1-Sprint-2-Statistics/blob/master/module3%5CDS_123_Confidence_Intervals_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Confidence Intervals

The following url can be used to access an abbreviated version of responses to Stack Overflow's 2018 Developer Survey. The original Survey had ~100k respondents but the data is quite dirty so I have selected a cleaner subset of it for you to use for your assignment.

<https://raw.githubusercontent.com/ryanleeallred/datasets/master/bootcampers.csv>

The provided dataset holds 14 columns of information about individuals who make less than 500k per year and who responded that they had: 

"Participated in a full-time developer training program or bootcamp"

## Part 1 - Setting the Stage

**1) Load the dataset**

In [5]:
import pandas as pd
import matplotlib.pyplot as pt
import numpy as np
from scipy import stats

todays_df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/bootcampers.csv')


In [6]:
print(todays_df.shape)
todays_df.head()

(2761, 15)


Unnamed: 0         Student  ...                Age                 RaceEthnicity
0          62              No  ...  25 - 34 years old  White or of European descent
1          73              No  ...  25 - 34 years old  White or of European descent
2         127  Yes, full-time  ...  25 - 34 years old                    East Asian
3         140              No  ...  25 - 34 years old  White or of European descent
4         153              No  ...  25 - 34 years old  White or of European descent

[5 rows x 15 columns]

**2) Select two random samples from this dataset, one with a sample size of 20 and the other with a sample size of 200. (Use a `random_state` of `42` when selecting the samples)**

In [21]:
first_random_sample = todays_df['ConvertedSalary'].sample(20, random_state=42)

first_random_sample.head()

367     74500.0
2759    60000.0
1330    86120.0
2750    60000.0
521     25047.0
Name: ConvertedSalary, dtype: float64

In [22]:
second_random_sample = todays_df['ConvertedSalary'].sample(200, random_state=42)

second_random_sample.head()

367     74500.0
2759    60000.0
1330    86120.0
2750    60000.0
521     25047.0
Name: ConvertedSalary, dtype: float64

**3) Calculate and report the sample means of the `ConvertedSalary` column for both of the samples.**

In [23]:
mean_20 = first_random_sample.mean()

mean_20

55752.2

In [24]:
mean_200 = second_random_sample.mean()

mean_200

68551.255

**4) Both of these sample means are estimates of an underlying population value. Which sample mean do you trust more? Why? Would a non-technical audience have any idea about which of these values is more trustworthy?**

In [27]:
def confidence_interval(sample, confidence_level=.95):
  
  # Turn this into a numpy array for consistency
  data = np.array(sample)
  
  # Sample Standard Deviation (ddof=1) with numpy to get sample metrics
  s = np.std(data, ddof=1)
  n = data.shape[0]
  x_bar = np.mean(data)
  
  # Look up the t statistic that corresponds to our sample size *and* confidence level
  t = stats.t.ppf(1 - ((1 - confidence_level) / 2), n-1)
  # Standard Error
  std_err = (s / np.sqrt(n))
  # Margin of Error
  moe = t*std_err

  lower_bound = x_bar - moe
  upper_bound = x_bar + moe

  return (lower_bound, x_bar, upper_bound, moe)

In [54]:
sample = todays_df.sample(20, random_state=42)

confidence_interval(sample['ConvertedSalary'])



(35805.10309625536, 55752.2, 75699.29690374463, 19947.096903744638)

In [52]:
sample = todays_df.sample(200, random_state=42)

confidence_interval(sample['ConvertedSalary'])

(60593.058804479086, 68551.255, 76509.45119552092, 7958.196195520917)

Based on a moe of + or - 7,958.196, I would trust the second sample more. The margin is smaller and someone from a non technical audience would trust it more based on that. 

**5) Does just the point estimate (individual value of the sample mean) indicate to you anything about how much sampling error there could be with these estimates?**

**6) What strategies could we use when reporting these numbers to not only report our estimates but also to give non-technical readers an idea about how far off our estimates might be due to sampling error?**



## Part 2 - Reporting CIs / MOEs along with our point estimates for more context.

**1) Calculate and report a 95% confidence interval around both of the sample means from part 1.**



**2) Which confidence interval is wider and why?**

The confidence interval of the first sample is much wider. Due to its lower sample size the standard error of our estimate is much larger causing the potential distribution of sample means to be much more spread out. A confidence interval catches 95% of this theoretical distriution of sample means so if our standard error is larger, our confidence interval will be wider as well.

**3) Report the mean and the margin of error for both of the sample means. What does the Margin of Error Represent?**

The margin of error gives an idea of how far off our estimates might be (with 95% confidence). We're trying to supply a plausible range for our parameter of interest (the true average salary of bootcamp grads).

## Part 3 - Communicate the Precision of Sample Estimates Graphically

**1) Create a plot using `plt.errorbar` that compares both of the confidence intervals.** 

**2) Create a plot using `plt.bar` that compares both of the confidence intervals.**

## Part 4 - Check for Understanding







**Calculate a Confidence Interval using the entire dataset. How precise do our estimates get?**

**What does "precision" mean in the context of statistical estimates and how is that different from "accuracy?"**

**It is very common to misunderstand what is captured by a 95% confidence interval. What is the correct interpretation? ([Hint](https://www.statisticssolutions.com/misconceptions-about-confidence-intervals/))**

# Stretch Goals:

1) Study the relationship between t-tests and confidence intervals.
 - Find a sample mean that we have worked with and construct a 95% confidence interval around it. (find the lower and upper bounds)
 - Run a 1-sample t-test with the null hypothesis value being just barely **outside** of the confidence interval. What is the p-value?
 - Run a 1-sample t-test with the null hypothesis value being just barely **inside** of the confidence interval. What is the p-value?

  What does it mean when we say that the boundaries of the confidence interval are the boundaries of statistical significance in a 1-sample t-test?


2) Go back to our [congressional voting dataset](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records) and build confidence intervals around the means of all of the issues for a single party. Compare all of those confidence intervals graphically on a single graph. 

3) Dive deeper into the [2018 Stack Overflow Survey](https://www.kaggle.com/stackoverflow/stack-overflow-2018-developer-survey) results to see what cool things you can find.